### PR TITLE
thread: use mach semaphores on osx

### PR DESCRIPTION
**Pull request by indutny**
nt

### CODE REVIEW

**Comment by piscisaureus** [comment]
No real objections, but: why?

**Comment by indutny** [comment]
That's a good question. AFAIK, because not every OSX version has support of posix semaphores (and I'm afraid that even on latest version support may be quite limited).

**Comment by indutny** [comment]
Better ask @bnoordhuis anyway

File: include/uv-private/uv-unix.h [hunk 1]
```diff
@@ -40,7 +40,14 @@
 #include <termios.h>
 #include <pwd.h>
 
+#if defined(__APPLE__) && defined(__MACH__)
+#include <mach/mach.h>
```
Comment (bnoordhuis) [inline include/uv-private/uv-unix.h:44]: `# include <...>`

File: src/unix/thread.c [hunk 2]
```diff
@@ -168,6 +168,45 @@ void uv_once(uv_once_t* guard, void (*callback)(void)) {
...
+    abort();
+}
+
+
+int uv_sem_trywait(uv_sem_t* sem) {
+  mach_timespec_t interval;
+  interval.tv_sec = 0;
+  interval.tv_nsec = 0;
+
+  /* A little bit hacky but should work */
```
Comment (bnoordhuis) [inline src/unix/thread.c:200]: You can remove the comment. I checked the xnu source - a zero timeout polls, it never blocks.

File: src/unix/thread.c [hunk 3]
```diff
@@ -168,6 +168,45 @@ void uv_once(uv_once_t* guard, void (*callback)(void)) {
 }
 
 
+#if defined(__APPLE__) && defined(__MACH__)
```
Comment (bnoordhuis) [inline src/unix/thread.c:171]: Add a newline here.

EDIT: Ah, never mind. I'll touch it up when I merge it.

**Comment by bnoordhuis** [comment]
Thanks Fedor, landed in 3a54805. I'm a bit unclear on how Mach semaphores interact with signals, i.e. if they're interruptible, but a quick scan of sync_sema.c suggests they're not.

**Comment by indutny** [comment]
Landed in 3a548053b8d758f04e1d15c84daadc49fb9ac69e

**Pull request closed without merging**